README.md
# Facility Management System - open source facility booking system for schools.

By Mark Edward Olango.

[![License: AGPLv3](https://img.shields.io/static/v1?label=License&message=AGPLv3&color=3DA639&style=flat-square)](https://www.gnu.org/licenses/agpl-3.0.html)
[![Twitter Follow](https://img.shields.io/twitter/follow/crbsapp.svg?style=social)](https://twitter.com/crbsapp)

This is a web-based facility booking system for schools and is designed to be as easy to use as possible. Set up your bookable rooms, day schedule and timetable for the year. Add user accounts, and allow them to make and manage bookings from anywhere.


It is web-based - PHP and MySQL - and currently uses the [CodeIgniter 3](https://codeigniter.com/) framework.

## Documentation

## Bug Reports & Feature Requests

## Security
To report any security issues, please email markedward.olango@ustp.edu.ph instead of using the issue tracker.

## Credits

Big Credits to David Craigway

This project makes use of several third parties, some of which are listed below.

- [CodeIgniter](https://codeigniter.com/) (MIT)
- [Unpoly](https://unpoly.com/) (MIT)
- [FamFamFam Silk Icons](http://www.famfamfam.com/lab/icons/silk/) ([CC BY 3.0](https://creativecommons.org/licenses/by/3.0/), Unmodified)


CHANGELOG.md
# Facility Management System Changelog


## [2.6.4] 2022-07-22

### Fixed
- Issue where a user with no department creating multiple single bookings would result in no bookings being created.


## [2.6.3] 2022-06-29

### Fixed
- Issue where Session calendar might not display the last month, depending on start/end date of Session.


## [2.6.2] 2022-03-26

### Fixed
- Bookings Grid display issues in Firefox - cells should now all be the same size.


## [2.6.1] 2022-03-02

### Fixed
- Issue when creating a recurring booking with no user selected.


## [2.6.0] 2022-01-29

### Added
- Sessions feature, to replace Academic Years. Create as many as you need and switch between them.
- Settings: Option to set timezone.
- Users page: Search filter.
- Dashboard: improved interface for teacher users.
- Bookings: Multiple selection mode.
- Bookings: Preview recurring bookings.
- Bookings: View detailed information for bookings.
- Bookings: View reasons when rooms/periods are unavailable.
- Bookings: Edit or Cancel individual instances of recurring bookings.
- Bookings: Ability for users to switch between Sessions.

### Changed
- Bookings section overhauled with many new features.
- Timetable Weeks: these are now configured within each Session.
- Holidays: These are now configured within each Session.
- Timetable Weeks: improved colour picker widget.

### Fixed
- Timetable weeks: now have a default colour, instead of none at all.


## [2.5.0] 2021-04-27

### Added
- Room Access Control feature - set visibility of each room to specific Users or Departments.

### Changed
- Updated sizing for Date Picker widget.
- Added cache-busting query strings to JS/CSS assets.

### Fixed
- Issue with loading override language entries from database.


## [2.4.1] 2020-12-09

### Fixed
- Potential SQL security issue when importing users from CSV file.


## [2.4.0] 2020-03-10

### Added
- Support for setting a custom message to appear on the login page.
- Support for controlling the visibility of booking user details to other users.

### Changed
- Improvements for back/next navigation between days; will now skip days that don't have any periods.


## [2.3.2] 2020-09-25

### Fixed
- Issue saving school details when an error occurs.


## [2.3.1] 2020-09-25

### Fixed
- Issue when saving school details when an error occurs.


## [2.3.0] 2020-08-26

### Added
- Support for LDAP authentication.
- Support for language line overrides in the database.

### Changed
- General javascript tidy-up and library updates.
- New style of room information popup on Bookings pgae.

### Fixed
- Removed erroneous debugging output from Weeks model.


## [2.2.0] 2020-06-06

### Added
- New settings page for additional settings.
- New setting for 'maximum active bookings': specify how many active bookings a user can have at one time.
- Date and Time display formats on Bookings page can now be customised.
- License details.

### Changed
- Updated icons to better quality PNG format.

### Fixed
- Issue where bookings on Sundays were not being displayed.


## [2.1.3] 2020-03-11

### Fixed
- Fixed another situation where existing bookings were being detected incorrectly.


## [2.1.2] 2019-12-03

### Fixed
- Fixed previous fix for database detection during installation which affected post-install.


## [2.1.1] 2019-12-02

### Fixed
- Fixed issue where Install page wouldn't load/would display errors when trying to load database.


## [2.1.0] 2019-11-13

Introducing the new 'Maintenance Mode' feature.

When enabled, Maintenance Mode prevents Teacher user accounts from viewing and making changes to bookings. The message can be customised, and will be displayed at the top of all pages.


### Added
- Added a new section to the School Details settings page to manage Maintenance Mode.

### Fixed
- Fixed an issue with one of the database migrations that might occur when updating from a pervious version.


## [2.0.5] 2019-09-30

Another update to "existing booking" check and minor tweaks.

### Changed
- Updated "Existing Bookings" check to make sure the Weekday was properly included.
- Updated the 'Recurring' section of the 'Make a booking' page to default the weekday value to the weekday of the chosen date.
- Updated 'Add week' page to have a default contrasting background colour.



## [2.0.4] 2019-08-31

Another update to make the "existing booking" check more robust.

### Changed
- Updated "Existing Bookings" check to make sure the Week was properly included.


## [2.0.3] 2019-08-31

Small update to address a Holiday display issue.

### Changed
- Updated Bookings to display Holiday details, when applicable, instead of an existing static or recurring booking.


## [2.0.2] 2019-06-19

One small bugfix.

### Changed
- Fixed an issue with User and Department lists having default limit applied when they shouldn't.


## [2.0.1] - 2019-01-26

Minor fix to day settings for periods and addition of favicon.

### Added
- Favicon to help facility management system stand out in tabs and windows.

### Changed
- Fixed an issue relating to possible issues with period days being shifted by one if upgraded from v1 to v2.


## [2.0.0] - 2019-01-02

The big one! Major update to support modern PHP, plus others.

### Added
- PHP Requirement for minimum version 5.5, and support for 7.x.
- Use Composer for dependencies.
- Database migrations.
- New installer.
- New upgrader for v1 => v2.

### Changed
- Updated CodeIgniter framework to version 3.
- Updated all class files for compatibility with CodeIgniter 3.
- Updated folder structure and configuration file methods.
- Security updates for HTML escaping.
- Fixed various bugs.
- Bitmask library for period/lesson time days.

### Removed
- Header image colour generation.


index2.php
<?php

$env = 'production';
if (is_file(__DIR__ . '/.env')) {
    $env_file = trim(file_get_contents(__DIR__ . '/.env'));
    $env = (empty($env_file) ? $env : $env_file);
}

define('ENVIRONMENT', isset($_SERVER['CI_ENV']) ? $_SERVER['CI_ENV'] : $env);

switch (ENVIRONMENT) {
    case 'development':
        error_reporting(-1);
        ini_set('display_errors', 1);
        break;

    case 'testing':
    case 'production':
        ini_set('display_errors', 0);
        if (version_compare(PHP_VERSION, '5.3', '>=')) {
            error_reporting(E_ALL & ~E_NOTICE & ~E_DEPRECATED & ~E_STRICT & ~E_USER_NOTICE & ~E_USER_DEPRECATED);
        } else {
            error_reporting(E_ALL & ~E_NOTICE & ~E_STRICT & ~E_USER_NOTICE);
        }
        break;

    default:
        header('HTTP/1.1 503 Service Unavailable.', TRUE, 503);
        echo 'The application environment is not set correctly.';
        exit(1);
}

$system_path = 'application/vendor/codeigniter/framework/system';

$application_folder = 'application';

$view_folder = '';

define('BASEPATH', $system_path . DIRECTORY_SEPARATOR);

define('APPPATH', $application_folder . DIRECTORY_SEPARATOR);

define('VIEWPATH', ($view_folder == '' ? APPPATH . 'views' . DIRECTORY_SEPARATOR : $view_folder . DIRECTORY_SEPARATOR));

require_once BASEPATH . 'core/CodeIgniter.php';


assets/js/main.js
// up.motion.config.duration = 150;
up.motion.config.enabled = false;
// up.history.config.enabled = false;


/**
 * Settings: Initialise sortable tables
 *
 */
up.compiler('.sort-table', function(element, cols) {
	new SortableTable(element, cols);
});


/**
 * Bookings: Automatically submit form when picking a date.
 *
 */
up.compiler('#bookings_date', function(element) {
	var input = document.getElementById("chosen_date"),
		form = document.getElementById("bookings_date");
	input.addEventListener("change", function(event) {
		form.submit();
	});
});


/**
 * Settings: LDAP: Form for testing the settings
 *
 */
up.compiler('[ldap-settings]', function(element) {

	function populateTestForm() {
		var attrs = up.Params.fromForm('#ldap_settings'),
			attrsArray = attrs.toArray();
		for (var i = 0; i < attrsArray.length; i++) {
			var sel = "[ldap-settings] [name='" + attrsArray[i]['name'] + "']";
			var dest = "[ldap-test] [name='" + attrsArray[i]['name'] + "']";
			var testEl = up.element.get(dest);
			if (testEl) {
				up.element.setAttrs(testEl, { value: attrsArray[i]['value'] });
			}
		}
	};

	populateTestForm();

	up.observe(element, { batch: true }, function(diff) {
		// console.log('Observed one or more changes: %o', diff)
		populateTestForm();
	});

});


/**
 * Session calendars: Config mode: Click on dates to change Timetable Week assignments.
 *
 */
up.compiler('.session-calendars.mode-config', function(sessionEl, data) {

	var weekIds = [];
	var weekClassNames = [];
	var numWeeks = data && data.weeks ? data.weeks.length : 0;

	// Process list of available weeks
	for (var i = 0; i < numWeeks; i++) {
		var weekId = parseInt(data.weeks[i], 10);
		className = 'week-' + weekId;
		weekIds.push(weekId);
		weekClassNames.push(className);
	}
	// Add blank entry for cycling through items
	weekIds.push(null);

	// Click event for changing week assignment of a given date
	up.on(sessionEl, 'click', '.date-btn', function(event, el) {

		var cell = el.closest('.date-cell');

		var weekstart = cell.getAttribute('data-weekstart');
		var weekid = cell.getAttribute('data-weekid');

		// Current Week ID
		var curWeekId = weekid ? '' + weekid : null;
		// Array index of the current selected week
		var curWeekIdx = weekIds.indexOf(parseInt(curWeekId, 10));
		// Array index of next week to select
		var nextWeekIdx = (curWeekIdx + 1) % weekIds.length;

		// Next Week ID from the list
		var newWeekId = weekIds[ (curWeekIdx + 1) % weekIds.length ];

		// All cells (potentially on other months) whose week starts on `weekstart`
		var cells = up.element.all(sessionEl, '.date-cell[data-weekstart="' + weekstart + '"]');

		// Process all cells that match start of week
		up.util.each(cells, function(cell) {

			// Remove existing week classes
			for (var i = 0; i < weekClassNames.length; i++) {
				cell.classList.remove(weekClassNames[i]);
			}

			// An actual Week ID to add
			if (newWeekId !== null) {
				cell.classList.add('week-' + newWeekId);
			}

			// Update value of hidden input
			var input = up.element.get(cell, 'input[type=hidden]');
			input.value = newWeekId;

			// Update the data- attr with the new Week ID
			cell.setAttribute('data-weekid', newWeekId);

		});

	});

});


up.compiler('.up-datepicker', function(el) {
	up.on(el, 'click', function(evt, el, data) {
		if ( ! data.input) return;
		return displayDatePicker(data.input, false);
	});
});


/**
 * Bookings page: controls forms
 *
 */
up.compiler('#bookings_controls_day', function(form) {

	// Date picker and form submission

	// up.on(form, 'click', '.up-datepicker', function(evt, el, data) {
	// 	if ( ! data.input) return;
	// 	return displayDatePicker(data.input, false);
	// });

	up.on(form, 'change', '.up-datepicker-input', function(evt, el, data) {
		form.submit();
		up.emit(form, 'submit');
	});

});

up.compiler('#bookings_controls_room', function(form) {

	// Room list

	up.on(form, 'change', 'select[name=room]', function(evt, el, data) {
		form.submit();
	});

});


up.compiler('#bookings_controls_session', function(form) {

	// Session list

	up.on(form, 'change', 'select[name=session_id]', function(evt, el, data) {
		form.submit();
	});

});


up.compiler('[up-copy-to]', function(copyBtn) {

	up.on(copyBtn, 'click', function(evt, el, data) {

		var group = el.getAttribute('up-copy-to');

		var srcEl = up.element.get('[up-copy-group="' + group + '"]');
		var value = srcEl.value;

		var allItems = up.element.all('[up-copy-group="' + group + '"]');
		up.util.each(allItems, function(item) {
			item.value = value;
		});

	});
});


up.compiler('select[up-autocomplete]', function(selectEl) {
	accessibleAutocomplete.enhanceSelectElement({
		selectElement: selectEl,
		displayMenu: 'overlay',
		dropdownArrow: function(obj) {
			return '<svg xmlns="http://www.w3.org/2000/svg" focusable="false" class="autocomplete__dropdown-arrow-down h-5 w-5" viewBox="0 0 20 20" fill="currentColor"><path fill-rule="evenodd" d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z" clip-rule="evenodd" />';
		}
	});
});



/**
 * Unavailable slots: popup content: change position of popup based on X pos of element.
 *
 * This is used for unavailable slots - like holidays or period unavailability.
 *
 */
up.macro('.bookings-grid-button[up-content]', function(el) {
	el.setAttribute('up-position', getPopupAlignment(el));
});


/**
 * Determine where to place the popup (attached via data-up-popup).
 *
 */
function getPopupAlignment(el) {

	var rect = el.getBoundingClientRect();
	var posX = Math.ceil(rect.x);
	var docWidth = document.documentElement.clientWidth;
	var docHalf = Math.floor(docWidth / 2);

	if (posX < docHalf) {
		return 'right';
	} else {
		return 'left';
	}
}


assets/js/lib/sorttable.js

/*----------------------------------------------------------------------------\
|                            Sortable Table 1.12                              |
|-----------------------------------------------------------------------------|
|                         Created by Erik Arvidsson                           |
|                  (http://webfx.eae.net/contact.html#erik)                   |
|                      For WebFX (http://webfx.eae.net/)                      |
|-----------------------------------------------------------------------------|
| A DOM 1 based script that allows an ordinary HTML table to be sortable.     |
|-----------------------------------------------------------------------------|
|                  Copyright (c) 1998 - 2004 Erik Arvidsson                   |
|-----------------------------------------------------------------------------|
| This software is provided "as is", without warranty of any kind, express or |
| implied, including  but not limited  to the warranties of  merchantability, |
| fitness for a particular purpose and noninfringement. In no event shall the |
| authors or  copyright  holders be  liable for any claim,  damages or  other |
| liability, whether  in an  action of  contract, tort  or otherwise, arising |
| from,  out of  or in  connection with  the software or  the  use  or  other |
| dealings in the software.                                                   |
| - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - |
| This  software is  available under the  three different licenses  mentioned |
| below.  To use this software you must chose, and qualify, for one of those. |
| - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - |
| The WebFX Non-Commercial License          http://webfx.eae.net/license.html |
| Permits  anyone the right to use the  software in a  non-commercial context |
| free of charge.                                                             |
| - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - |
| The WebFX Commercial license           http://webfx.eae.net/commercial.html |
| Permits the  license holder the right to use  the software in a  commercial |
| context. Such license must be specifically obtained, however it's valid for |
| any number of  implementations of the licensed software.                    |
| - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - |
| GPL - The GNU General Public License    http://www.gnu.org/licenses/gpl.txt |
| Permits anyone the right to use and modify the software without limitations |
| as long as proper  credits are given  and the original  and modified source |
| code are included. Requires  that the final product, software derivate from |
| the original  source or any  software  utilizing a GPL  component, such  as |
| this, is also licensed under the GPL license.                               |
|-----------------------------------------------------------------------------|
| 2003-01-10 | First version                                                  |
| 2003-01-19 | Minor changes to the date parsing                              |
| 2003-01-28 | JScript 5.0 fixes (no support for 'in' operator)               |
| 2003-02-01 | Sloppy typo like error fixed in getInnerText                   |
| 2003-07-04 | Added workaround for IE cellIndex bug.                         |
| 2003-11-09 | The bDescending argument to sort was not correctly working     |
|            | Using onclick DOM0 event if no support for addEventListener    |
|            | or attachEvent                                                 |
| 2004-01-13 | Adding addSortType and removeSortType which makes it a lot     |
|            | easier to add new, custom sort types.                          |
| 2004-01-27 | Switch to use descending = false as the default sort order.    |
|            | Change defaultDescending to suit your needs.                   |
| 2004-03-14 | Improved sort type None look and feel a bit                    |
| 2004-08-26 | Made the handling of tBody and tHead more flexible. Now you    |
|            | can use another tHead or no tHead, and you can chose some      |
|            | other tBody.                                                   |
|-----------------------------------------------------------------------------|
| Created 2003-01-10 | All changes are in the log above. | Updated 2004-08-26 |
\----------------------------------------------------------------------------*/


function SortableTable(oTable, oSortTypes) {

	this.sortTypes = oSortTypes || [];

	this.sortColumn = null;
	this.descending = null;

	var oThis = this;
	this._headerOnclick = function (e) {
		oThis.headerOnclick(e);
	};

	if (oTable) {
		this.setTable( oTable );
		this.document = oTable.ownerDocument || oTable.document;
	}
	else {
		this.document = document;
	}


	// only IE needs this
	var win = this.document.defaultView || this.document.parentWindow;
	this._onunload = function () {
		oThis.destroy();
	};
	if (win && typeof win.attachEvent != "undefined") {
		win.attachEvent("onunload", this._onunload);
	}
}

SortableTable.gecko = navigator.product == "Gecko";
SortableTable.msie = /msie/i.test(navigator.userAgent);
// Mozilla is faster when doing the DOM manipulations on
// an orphaned element. MSIE is not
SortableTable.removeBeforeSort = SortableTable.gecko;

SortableTable.prototype.onsort = function () {};

// default sort order. true -> descending, false -> ascending
SortableTable.prototype.defaultDescending = false;

// shared between all instances. This is intentional to allow external files
// to modify the prototype
SortableTable.prototype._sortTypeInfo = {};

SortableTable.prototype.setTable = function (oTable) {
	if ( this.tHead )
		this.uninitHeader();
	this.element = oTable;
	this.setTHead( oTable.tHead );
	this.setTBody( oTable.tBodies[0] );
};

SortableTable.prototype.setTHead = function (oTHead) {
	if (this.tHead && this.tHead != oTHead )
		this.uninitHeader();
	this.tHead = oTHead;
	this.initHeader( this.sortTypes );
};

SortableTable.prototype.setTBody = function (oTBody) {
	this.tBody = oTBody;
};

SortableTable.prototype.setSortTypes = function ( oSortTypes ) {
	if ( this.tHead )
		this.uninitHeader();
	this.sortTypes = oSortTypes || [];
	if ( this.tHead )
		this.initHeader( this.sortTypes );
};

// adds arrow containers and events
// also binds sort type to the header cells so that reordering columns does
// not break the sort types
SortableTable.prototype.initHeader = function (oSortTypes) {
	if (!this.tHead) return;
	var cells = this.tHead.rows[0].cells;
	var doc = this.tHead.ownerDocument || this.tHead.document;
	this.sortTypes = oSortTypes || [];
	var l = cells.length;
	var img, c;
	for (var i = 0; i < l; i++) {
		c = cells[i];
		if (this.sortTypes[i] != null && this.sortTypes[i] != "None") {
			img = doc.createElement("IMG");
			img.src = window.BASE_URL + "assets/images/blank.png";
			c.appendChild(img);
			if (this.sortTypes[i] != null)
				c._sortType = this.sortTypes[i];
			if (typeof c.addEventListener != "undefined")
				c.addEventListener("click", this._headerOnclick, false);
			else if (typeof c.attachEvent != "undefined")
				c.attachEvent("onclick", this._headerOnclick);
			else
				c.onclick = this._headerOnclick;
		}
		else
		{
			c.setAttribute( "_sortType", oSortTypes[i] );
			c._sortType = "None";
		}
	}
	this.updateHeaderArrows();
};

// remove arrows and events
SortableTable.prototype.uninitHeader = function () {
	if (!this.tHead) return;
	var cells = this.tHead.rows[0].cells;
	var l = cells.length;
	var c;
	for (var i = 0; i < l; i++) {
		c = cells[i];
		if (c._sortType != null && c._sortType != "None") {
			c.removeChild(c.lastChild);
			if (typeof c.removeEventListener != "undefined")
				c.removeEventListener("click", this._headerOnclick, false);
			else if (typeof c.detachEvent != "undefined")
				c.detachEvent("onclick", this._headerOnclick);
			c._sortType = null;
			c.removeAttribute( "_sortType" );
		}
	}
};

SortableTable.prototype.updateHeaderArrows = function () {
	if (!this.tHead) return;
	var cells = this.tHead.rows[0].cells;
	var l = cells.length;
	var img;
	for (var i = 0; i < l; i++) {
		if (cells[i]._sortType != null && cells[i]._sortType != "None") {
			img = cells[i].lastChild;
			if (i == this.sortColumn)
				img.className = "sort-arrow " + (this.descending ? "descending" : "ascending");
			else
				img.className = "sort-arrow";
		}
	}
};

SortableTable.prototype.headerOnclick = function (e) {
	// find TD element
	var el = e.target || e.srcElement;
	while (el.tagName != "TD")
		el = el.parentNode;

	this.sort(SortableTable.msie ? SortableTable.getCellIndex(el) : el.cellIndex);
};

// IE returns wrong cellIndex when columns are hidden
SortableTable.getCellIndex = function (oTd) {
	var cells = oTd.parentNode.childNodes
	var l = cells.length;
	var i;
	for (i = 0; cells[i] != oTd && i < l; i++)
		;
	return i;
};

SortableTable.prototype.getSortType = function (nColumn) {
	return this.sortTypes[nColumn] || "String";
};

// only nColumn is required
// if bDescending is left out the old value is taken into account
// if sSortType is left out the sort type is found from the sortTypes array

SortableTable.prototype.sort = function (nColumn, bDescending, sSortType) {
	if (!this.tBody) return;
	if (sSortType == null)
		sSortType = this.getSortType(nColumn);

	// exit if None
	if (sSortType == "None")
		return;

	if (bDescending == null) {
		if (this.sortColumn != nColumn)
			this.descending = this.defaultDescending;
		else
			this.descending = !this.descending;
	}
	else
		this.descending = bDescending;

	this.sortColumn = nColumn;

	if (typeof this.onbeforesort == "function")
		this.onbeforesort();

	var f = this.getSortFunction(sSortType, nColumn);
	var a = this.getCache(sSortType, nColumn);
	var tBody = this.tBody;

	a.sort(f);

	if (this.descending)
		a.reverse();

	if (SortableTable.removeBeforeSort) {
		// remove from doc
		var nextSibling = tBody.nextSibling;
		var p = tBody.parentNode;
		p.removeChild(tBody);
	}

	// insert in the new order
	var l = a.length;
	for (var i = 0; i < l; i++)
		tBody.appendChild(a[i].element);

	if (SortableTable.removeBeforeSort) {
		// insert into doc
		p.insertBefore(tBody, nextSibling);
	}

	this.updateHeaderArrows();

	this.destroyCache(a);

	if (typeof this.onsort == "function")
		this.onsort();
};

SortableTable.prototype.asyncSort = function (nColumn, bDescending, sSortType) {
	var oThis = this;
	this._asyncsort = function () {
		oThis.sort(nColumn, bDescending, sSortType);
	};
	window.setTimeout(this._asyncsort, 1);
};

SortableTable.prototype.getCache = function (sType, nColumn) {
	if (!this.tBody) return [];
	var rows = this.tBody.rows;
	var l = rows.length;
	var a = new Array(l);
	var r;
	for (var i = 0; i < l; i++) {
		r = rows[i];
		a[i] = {
			value:		this.getRowValue(r, sType, nColumn),
			element:	r
		};
	};
	return a;
};

SortableTable.prototype.destroyCache = function (oArray) {
	var l = oArray.length;
	for (var i = 0; i < l; i++) {
		oArray[i].value = null;
		oArray[i].element = null;
		oArray[i] = null;
	}
};

SortableTable.prototype.getRowValue = function (oRow, sType, nColumn) {
	// if we have defined a custom getRowValue use that
	if (this._sortTypeInfo[sType] && this._sortTypeInfo[sType].getRowValue)
		return this._sortTypeInfo[sType].getRowValue(oRow, nColumn);

	var s;
	var c = oRow.cells[nColumn];
	if (typeof c.innerText != "undefined")
		s = c.innerText;
	else
		s = SortableTable.getInnerText(c);
	return this.getValueFromString(s, sType);
};

SortableTable.getInnerText = function (oNode) {
	var s = "";
	var cs = oNode.childNodes;
	var l = cs.length;
	for (var i = 0; i < l; i++) {
		switch (cs[i].nodeType) {
			case 1: //ELEMENT_NODE
				s += SortableTable.getInnerText(cs[i]);
				break;
			case 3:	//TEXT_NODE
				s += cs[i].nodeValue;
				break;
		}
	}
	return s;
};

SortableTable.prototype.getValueFromString = function (sText, sType) {
	if (this._sortTypeInfo[sType])
		return this._sortTypeInfo[sType].getValueFromString( sText );
	return sText;
	/*
	switch (sType) {
		case "Number":
			return Number(sText);
		case "CaseInsensitiveString":
			return sText.toUpperCase();
		case "Date":
			var parts = sText.split("-");
			var d = new Date(0);
			d.setFullYear(parts[0]);
			d.setDate(parts[2]);
			d.setMonth(parts[1] - 1);
			return d.valueOf();
	}
	return sText;
	*/
	};

SortableTable.prototype.getSortFunction = function (sType, nColumn) {
	if (this._sortTypeInfo[sType])
		return this._sortTypeInfo[sType].compare;
	return SortableTable.basicCompare;
};

SortableTable.prototype.destroy = function () {
	this.uninitHeader();
	var win = this.document.parentWindow;
	if (win && typeof win.detachEvent != "undefined") {	// only IE needs this
		win.detachEvent("onunload", this._onunload);
	}
	this._onunload = null;
	this.element = null;
	this.tHead = null;
	this.tBody = null;
	this.document = null;
	this._headerOnclick = null;
	this.sortTypes = null;
	this._asyncsort = null;
	this.onsort = null;
};

// Adds a sort type to all instance of SortableTable
// sType : String - the identifier of the sort type
// fGetValueFromString : function ( s : string ) : T - A function that takes a
//    string and casts it to a desired format. If left out the string is just
//    returned
// fCompareFunction : function ( n1 : T, n2 : T ) : Number - A normal JS sort
//    compare function. Takes two values and compares them. If left out less than,
//    <, compare is used
// fGetRowValue : function( oRow : HTMLTRElement, nColumn : int ) : T - A function
//    that takes the row and the column index and returns the value used to compare.
//    If left out then the innerText is first taken for the cell and then the
//    fGetValueFromString is used to convert that string the desired value and type

SortableTable.prototype.addSortType = function (sType, fGetValueFromString, fCompareFunction, fGetRowValue) {
	this._sortTypeInfo[sType] = {
		type:				sType,
		getValueFromString:	fGetValueFromString || SortableTable.idFunction,
		compare:			fCompareFunction || SortableTable.basicCompare,
		getRowValue:		fGetRowValue
	};
};

// this removes the sort type from all instances of SortableTable
SortableTable.prototype.removeSortType = function (sType) {
	delete this._sortTypeInfo[sType];
};

SortableTable.basicCompare = function compare(n1, n2) {
	if (n1.value < n2.value)
		return -1;
	if (n2.value < n1.value)
		return 1;
	return 0;
};

SortableTable.idFunction = function (x) {
	return x;
};

SortableTable.toUpperCase = function (s) {
	return s.toUpperCase();
};

SortableTable.toDate = function (s) {
	var parts = s.split("-");
	var d = new Date(0);
	d.setFullYear(parts[0]);
	d.setDate(parts[2]);
	d.setMonth(parts[1] - 1);
	return d.valueOf();
};


// add sort types
SortableTable.prototype.addSortType("Number", Number);
SortableTable.prototype.addSortType("CaseInsensitiveString", SortableTable.toUpperCase);
SortableTable.prototype.addSortType("Date", SortableTable.toDate);
SortableTable.prototype.addSortType("String");
// None is a special case
